File: sys/codegen/lib/petscOptionsInsertString/petscOptionsInsertString.c
#include "petscOptionsInsertString.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(int varargin_3);
static void m2c_error(void);
static void b_m2c_error(int varargin_3)
{
  const char * msgid;
  const char * fmt;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsInsertString returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char * msgid;
  const char * fmt;
  msgid = "Petsc4m:petscOptionsInsertString:InputError";
  fmt = "The argument must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void emxInitArray_char_T(emxArray_char_T **pEmxArray, int numDimensions)
{
  emxInit_char_T(pEmxArray, numDimensions);
}

void petscOptionsInsertString(const emxArray_char_T *in_str, int *errCode,
  boolean_T *toplevel)
{
  PetscOptions obj;
  *toplevel = true;
  if ((!(in_str->size[1] == 0)) && (in_str->data[in_str->size[1] - 1] != '\x00'))
  {
    m2c_error();
  }

  obj = NULL;
  *errCode = PetscOptionsInsertString(obj, &in_str->data[0]);
  if (*errCode != 0) {
    b_m2c_error(*errCode);
  }
}

void petscOptionsInsertString_initialize(void)
{
}

void petscOptionsInsertString_terminate(void)
{
}

File: sys/petscOptionsInsertString.m
function [errCode, toplevel] = petscOptionsInsertString(in_str)
%Inserts options into the database from a string
%   errCode = petscOptionsInsertString(in_str)
%
%   in_str -string that contains options separated by blanks and
%           must be null-terminated
%   errCode(int) return code (0 indicates OK)
%
% SEE ALSO: petscOptionsInsertFile
%
% PETSc C interface:
%   PetscErrorCode  PetscOptionsInsertString(PetscOptions options,const char in_str[])
% http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscOptionsInsertString.html

%#codegen -args {coder.typeof(char(0), [1, inf])}

errCode = int32(-1);

if ~coder.target('MATLAB')
    toplevel = nargout>1;
    if ~isempty(in_str) && in_str(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsInsertString:InputError', ...
            'The argument must be a null-terminated string.')
    end

    errCode = coder.ceval('PetscOptionsInsertString', PETSC_NULL_OPTIONS, coder.rref(in_str));

    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsInsertString returned error code %d\n', errCode)
    end
end

end
